File: python/1074_元素和为目标值的子矩阵数量_hard.py
class Solution:
    def numSubmatrixSumTarget(self, matrix: List[List[int]], target: int) -> int:
        def subArrayFindK(array, k):
            cnt, pre, ans = Counter([0]), 0, 0
            for n in array:
                pre += n
                if pre-k in cnt:
                    ans += cnt[pre-k]
                cnt[pre] += 1
            return ans
        
        m, n = len(matrix), len(matrix[0])
        res = 0
        for i in range(m):
            data = [0] * n
            for j in range(i, m):
                for c in range(0, n):
                    data[c] += matrix[j][c]
                res += subArrayFindK(data, target)
        return res

File: cpp/1074_元素和为目标值的子矩阵数量_hard.cpp
class Solution {
public:
    int numSubmatrixSumTarget(vector<vector<int>>& matrix, int target) {
        int m = matrix.size(), n = matrix[0].size();
        int ans = 0;
        for (int i = 0; i < m; i++) {
            vector<int> data(n);
            for (int j = i; j < m; j++) {
                for (int c = 0; c < n; c++)
                    data[c] += matrix[j][c];
                ans += subArray(data, target);
            }
        }
        return ans;
    }
    int subArray(vector<int> &matrix, int k) {
        unordered_map<int, int> cnt;
        cnt[0] = 1;
        int pre = 0, ans = 0;
        for (auto m : matrix) {
            pre += m;
            if (cnt.find(pre-k) != cnt.end())
                ans += cnt[pre-k];
            cnt[pre]++;
        }
        return ans;
    }
};